lib/ipmi_backend_utils.pm

```diff
@@ -390,4 +395,84 @@ sub ipmitool {
+
+Enable SEV in the kernel, because it is disabled by default. This is done by putting
+the following onto the kernel command line: mem_encrypt=on kvm_amd.sev=1. To make the 
+changes persistent, append the above to the variable holding parameters of the kernel
+command line in /etc/default/grub to preserve SEV settings across reboots:
+$ cat /etc/default/grub
+...
+GRUB_CMDLINE_LINUX="... mem_encrypt=on kvm_amd.sev=1"
+...
+mem_encrypt=on turns on the SME memory encryption feature on the host which protects 
+against the physical attack on the hypervisor memory. The kvm_amd.sev parameter 
+actually enables SEV in the kvm module. This subroutine receives only two arguments,
+the dst_machine is the host on which operations will be performed, the root_dir is 
+the partition on which grub files reside. If these two arguments are not given any
+values, operations will be performed on localhost and '/' partition. This subroutine
+calls add_kernel_options to do the actual kernel options adding work.
+
+=cut
+sub enable_sev_in_kernel {
+    my (%args) = @_;
+
+    $args{dst_machine} //= 'localhost';
+    $args{root_dir} //= '';
+    $args{root_dir} .= '/' unless $args{root_dir} =~ /\/$/;
+    croak("No AMD EPYC cpu on $args{dst_machine}, so sev can not be enabled in kernel.") unless (script_run("lscpu | grep -i \'AMD EPYC\'") == 0);
+    add_kernel_options(dst_machine => $args{dst_machine}, root_dir => $args{root_dir}, kernel_opts => 'mem_encrypt=on kvm_amd.sev=1');
+}
+
+=head2 add_kernel_options
+
+  add_kernel_options(dst_machine => $dst_machine, root_dir => $root_dir, kernel_opts => $options)
+
+Adding additional kernel options onto kernel command line in grub config file and 
+also GRUB_CMDLINE_LINUX_DEFAULT line in default grub config file. This subroutine 
+receives only three arguments, dst_machine is the host on which operations will be 
+performed, root_dir is the partition on which grub files reside, kernel_opts holds 
+a single text string this is composed of kernel options separated by spaces. If no
+values are passed in to dst_machine and root_dir, operations will be performed on 
+localhost and '/' partition by default.  Almost all regular kernel options can be
+passed in directly without modification except for very extreme cases in which very
+special characters should be treated specially and even escaped before being used.
+
+=cut
+sub add_kernel_options {
+    my (%args) = @_;
+
+    $args{dst_machine} //= 'localhost';
+    $args{root_dir} //= '';
+    $args{root_dir} .= '/' unless $args{root_dir} =~ /\/$/;
+    $args{kernel_opts} //= '';
+    croak("Nothing to added onto kernel command line. Argument kernel_opts should not be empty.") if ($args{kernel_opts} eq '');
+
+    my @options = split(/ /, $args{kernel_opts});
+    my $cmd = '';
+    my $grub_cfg_file = "$args{root_dir}boot/grub2/grub.cfg";
+    foreach (@options) {
+        $cmd = "sed -i -r \'s/$_//g; /^\\s{1,}linux\\s{1,}\\\/boot\\\// s/\$/ $_/g\' $grub_cfg_file";
```

Comment:
Would you please explain what this sed is doing?

Reply:
It search the linux kernel line in $grub_cfg_file, then replace and append new options onto its command line.